Add `GMNConv` and an example on `ZINC`

asarigun opened the pull request:
Adding The Graph Mixer Networks (https://arxiv.org/abs/2301.12493) Conv operator and example on ZINC

asarigun reviewed (commented):
I hope you find the implementation fully functional for PyG!

asarigun reviewed (commented):
Reviewed the changes for `GMNConv`

dufourc1 reviewed (commented):
Hi! Thanks for the implementation, I'll read the paper linked to make sure I'm aligned, you'll probably just need to add some tests.

dufourc1 reviewed (commented):
Hi, sorry for the delay and thanks again for the implementation ! Here are a few notes on the code and some more general comments.

Parts of the method are similar to PNA ([`PNAConv`](https://github.com/pyg-team/pytorch_geometric/blob/master/torch_geometric/nn/conv/pna_conv.py)). I wonder if there is a way of reusing the code already available ? What do you think ? 

For the tests you might want to check what is done here https://github.com/pyg-team/pytorch_geometric/blob/master/test/nn/conv/test_pna_conv.py.

asarigun commented on torch_geometric/nn/conv/gmn_conv.py:
```diff
@@ -0,0 +1,224 @@
+from torch_geometric.typing import Adj, OptTensor
+from torch_geometric.utils import degree, scatter
+
+from ..inits import reset
```
Since it will be inherited from the `PNAConv`, it's better to remove it!

asarigun commented on torch_geometric/nn/conv/gmn_conv.py:
```diff
@@ -0,0 +1,224 @@
+                  dim_size: Optional[int] = None) -> Tensor:
+
+        outs = []
+        for aggregator in self.aggregators:
+            if aggregator == 'sum':
+                out = scatter(inputs, index, 0, dim_size, reduce='sum')
+            elif aggregator == 'mean':
+                out = scatter(inputs, index, 0, dim_size, reduce='mean')
+            elif aggregator == 'min':
+                out = scatter(inputs, index, 0, dim_size, reduce='min')
+            elif aggregator == 'max':
+                out = scatter(inputs, index, 0, dim_size, reduce='max')
+            elif aggregator == 'var' or aggregator == 'std':
+                mean = scatter(inputs, index, 0, dim_size, reduce='mean')
+                mean_squares = scatter(inputs * inputs, index, 0, None,
+                                       dim_size, reduce='mean')
+                out = mean_squares - mean * mean
+                if aggregator == 'std':
+                    out = torch.sqrt(torch.relu(out) + 1e-5)
+            else:
+                raise ValueError(f'Unknown aggregator "{aggregator}".')
+            outs.append(out)
```
Since again, it will be inherited from the `PNAConv`, it's better to remove it!

asarigun commented on torch_geometric/nn/conv/gmn_conv.py:
```diff
@@ -0,0 +1,224 @@
+        deg = deg.clamp_(1).view(-1, 1, 1)
+
+        outs = []
+        for scaler in self.scalers:
+            if scaler == 'identity':
+                pass
+            elif scaler == 'amplification':
+                out = out * (torch.log(deg + 1) / self.avg_deg['log'])
+            elif scaler == 'attenuation':
+                out = out * (self.avg_deg['log'] / torch.log(deg + 1))
+            elif scaler == 'linear':
+                out = out * (deg / self.avg_deg['lin'])
+            elif scaler == 'inverse_linear':
+                out = out * (self.avg_deg['lin'] / deg)
+            else:
+                raise ValueError(f'Unknown scaler "{scaler}".')
```
Sure, it would be LGTM!

asarigun commented on torch_geometric/nn/conv/gmn_conv.py:
```diff
@@ -0,0 +1,224 @@
+            outs.append(out)
+        return torch.cat(outs, dim=-1)
+
+    def forward(self, x: Tensor, edge_index: Adj,
+                edge_attr: OptTensor = None) -> Tensor:
+
+        if self.divide_input:
+            x = x.view(-1, self.towers, self.F_in)
+        else:
+            x = x.view(-1, 1, self.F_in).repeat(1, self.towers, 1)
+
+        # propagate_type: (x: Tensor, edge_attr: OptTensor)
+        out = self.propagate(edge_index, x=x, edge_attr=edge_attr, size=None)
+
+        out = torch.cat([x, out], dim=-1)
+        outs = [nn(out[:, i]) for i, nn in enumerate(self.mix)]
+        out = torch.cat(outs, dim=1)
```
You're right! i will just add `message` function and remove this.

asarigun commented on examples/gmn.py:
```diff
@@ -0,0 +1,145 @@
+val_loader = DataLoader(val_dataset, batch_size=64)
+test_loader = DataLoader(test_dataset, batch_size=64)
+
+# Compute in-degree histogram over training data.
+deg = torch.zeros(5, dtype=torch.long)
+for data in train_dataset:
+    d = degree(data.edge_index[1], num_nodes=data.num_nodes, dtype=torch.long)
+    deg += torch.bincount(d, minlength=deg.numel())
```
Sure! I will change it!

asarigun commented:
Hi! Thanks for the feedback! I rearrange the code that you mentioned in the comments. 



> Hi, sorry for the delay and thanks again for the implementation ! Here are a few notes on the code and some more general comments.
> 
> Parts of the method are similar to PNA ([`PNAConv`](https://github.com/pyg-team/pytorch_geometric/blob/master/torch_geometric/nn/conv/pna_conv.py)). I wonder if there is a way of reusing the code already available ? What do you think ?
> 
> For the tests you might want to check what is done here https://github.com/pyg-team/pytorch_geometric/blob/master/test/nn/conv/test_pna_conv.py.

dufourc1 commented on torch_geometric/nn/conv/gmn_conv.py:
```diff
@@ -0,0 +1,150 @@
+        self.reset_parameters()
+
+    def message(self, x_j: Tensor) -> Tensor:
+        return x_j
```
I'm not too sure, the GMN should be similar to PNA in its first step, right ? Shouldn't this then  be similar to https://github.com/pyg-team/pytorch_geometric/blob/89f4873b411371880fb64344615ee9babef67d5e/torch_geometric/nn/conv/pna_conv.py#L178-L188 ?

dufourc1 commented on torch_geometric/nn/conv/gmn_conv.py:
```diff
@@ -0,0 +1,150 @@
+        super().__init__(in_channels, out_channels, aggregators, scalers, deg, 
+                         edge_dim, towers, divide_input, **kwargs)
+
+        if divide_input:
+            assert in_channels % towers == 0
+        assert out_channels % towers == 0
+
+        self.in_channels = in_channels
+        self.out_channels = out_channels
+        self.aggregators = aggregators
+        self.scalers = scalers
+        self.edge_dim = edge_dim
+        self.towers = towers
+        self.divide_input = divide_input
+
+        self.F_in = in_channels // towers if divide_input else in_channels
+        self.F_out = self.out_channels // towers
+
+        if self.edge_dim is not None:
+            self.edge_encoder = Linear(edge_dim, self.F_in)
```
These are already defined with line 104, you should be able to just reuse them ? https://github.com/pyg-team/pytorch_geometric/blob/89f4873b411371880fb64344615ee9babef67d5e/torch_geometric/nn/conv/pna_conv.py#L112-L126

asarigun replied:
Certainly, that approach works. I'll make the updates as you've suggested.

dufourc1 commented on test/nn/conv/test_gmn_conv.py:
```diff
@@ -0,0 +1,78 @@
+        assert torch.allclose(jit(x, adj.t()), out, atol=1e-6)
+
+
+@onlyNeighborSampler
+def test_gmn_conv_get_degree_histogram_neighbor_loader():
+    edge_index = torch.tensor([[0, 0, 0, 1, 1, 2, 3], [1, 2, 3, 2, 0, 0, 0]])
+    data = Data(num_nodes=5, edge_index=edge_index)
+    loader = NeighborLoader(
+        data,
+        num_neighbors=[-1],
+        input_nodes=None,
+        batch_size=5,
+        shuffle=False,
+    )
+    deg_hist = GMNConv.get_degree_histogram(loader)
+    assert torch.equal(deg_hist, torch.tensor([1, 2, 1, 1]))
+
+
+def test_gmn_conv_get_degree_histogram_dataloader():
+    edge_index_1 = torch.tensor([[0, 0, 0, 1, 1, 2, 3], [1, 2, 3, 2, 0, 0, 0]])
+    edge_index_2 = torch.tensor([[1, 1, 2, 2, 0, 3, 3], [2, 3, 3, 1, 1, 0, 2]])
+    edge_index_3 = torch.tensor([[1, 3, 2, 0, 0, 4, 2], [2, 0, 4, 1, 1, 0, 3]])
+    edge_index_4 = torch.tensor([[0, 1, 2, 4, 0, 1, 3], [2, 3, 3, 1, 1, 0, 2]])
+
+    data_1 = Data(num_nodes=5, edge_index=edge_index_1)  # hist = [1, 2 ,1 ,1]
+    data_2 = Data(num_nodes=5, edge_index=edge_index_2)  # hist = [1, 1, 3]
+    data_3 = Data(num_nodes=5, edge_index=edge_index_3)  # hist = [0, 3, 2]
+    data_4 = Data(num_nodes=5, edge_index=edge_index_4)  # hist = [1, 1, 3]
+
+    loader = DataLoader(
+        [data_1, data_2, data_3, data_4],
+        batch_size=1,
+        shuffle=False,
+    )
+    deg_hist = GMNConv.get_degree_histogram(loader)
+    assert torch.equal(deg_hist, torch.tensor([3, 7, 9, 1]))
```
Duplicate of https://github.com/pyg-team/pytorch_geometric/blob/89f4873b411371880fb64344615ee9babef67d5e/test/nn/conv/test_pna_conv.py#L46-L78

This just retest the `PNAConv` functions

asarigun replied:
Should I refactor that section of the code?

dufourc1 replied:
I think you can just remove it if it doesn't test something specific to your method

dufourc1 commented on torch_geometric/nn/conv/gmn_conv.py:
```diff
@@ -0,0 +1,150 @@
+        if self.edge_dim is not None:
+            self.edge_encoder = Linear(edge_dim, self.F_in)
+
+        self.mix = ModuleList()
```
Couldn't you just replace `self.mix` with `self.post_nns` ?

asarigun replied:
Certainly, that approach works. I'll make the updates as you've suggested.

dufourc1 reviewed (commented):
Sorry for the delay, thanks a lot for the requested changes 👍🏽 

I have added a few comments, but maybe it's only because I am confused with the similarity with `PNAConv`.

asarigun commented:
Hello @dufourc1,

Thank you for your insights on the code. I have made updates in the sections you highlighted and added comments where you've provided feedback. 

> Sorry for the delay, thanks a lot for the requested changes 👍🏽
> 
> I have added a few comments, but maybe it's only because I am confused with the similarity with `PNAConv`.

dufourc1 reviewed (changes requested):
Hi ! The main points look good to me. Once you've updated your code to pass the CI tests, 
this should be all good.

asarigun commented:
Thank you so much! I did something wrong unintentionally, how can I reopen the PR? 

> Hi ! The main points look good to me. Once you've updated your code to pass the CI tests, this should be all good.

asarigun commented:
Reopen!

asarigun commented:
Thank you so much! I did something wrong unintentionally, how can I reopen the PR?

> Hi ! The main points look good to me. Once you've updated your code to pass the CI tests, this should be all good.

asarigun committed:
add `GMN`

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

asarigun committed:
`update`

asarigun committed:
Merge branch 'master' of https://github.com/asarigun/pytorch_geometric

asarigun committed:
`update`

asarigun committed:
`update`

asarigun committed:
`update`

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

asarigun committed:
`update`

asarigun committed:
Merge branch 'master' of https://github.com/asarigun/pytorch_geometric

asarigun committed:
`update`

asarigun committed:
`update`

asarigun committed:
`update`

asarigun committed:
`update`

asarigun committed:
`update`

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

asarigun committed:
`update`

asarigun committed:
Merge branch 'master' of https://github.com/asarigun/pytorch_geometric

asarigun committed:
`update`

asarigun committed:
`update`

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

asarigun committed:
`update`

asarigun committed:
`edit`

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

asarigun committed:
`update`

asarigun committed:
Merge branch 'master' of https://github.com/asarigun/pytorch_geometric

asarigun committed:
`update`

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

asarigun committed:
`update`

asarigun committed:
Merge branch 'master' of https://github.com/asarigun/pytorch_geometric

asarigun committed:
`update`

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

asarigun commented:
Hi @dufourc1,

The code has successfully passed the CI checks! Thank you once again for your feedback🙌.
